File: src/brs/web/api/http/handler/GetBlockchainStatus.java
package brs.web.api.http.handler;

import brs.Block;
import brs.Blockchain;
import brs.BlockchainProcessor;
import brs.Burst;
import brs.peer.Peer;
import brs.props.Props;
import brs.services.TimeService;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.LegacyDocTag;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.ResultFields.*;

public final class GetBlockchainStatus extends ApiServlet.JsonRequestHandler {

  private final BlockchainProcessor blockchainProcessor;
  private final Blockchain blockchain;
  private final TimeService timeService;

  public GetBlockchainStatus(BlockchainProcessor blockchainProcessor, Blockchain blockchain, TimeService timeService) {
    super(new LegacyDocTag[]{LegacyDocTag.BLOCKS, LegacyDocTag.INFO});
    this.blockchainProcessor = blockchainProcessor;
    this.blockchain = blockchain;
    this.timeService = timeService;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) {
    JsonObject response = new JsonObject();
    response.addProperty("application", Burst.getPropertyService().getString(Props.APPLICATION));
    response.addProperty("version", Burst.getPropertyService().getString(Props.VERSION));
    response.addProperty(TIME_RESPONSE, timeService.getEpochTime());
    Block lastBlock = blockchain.getLastBlock();
    response.addProperty("lastBlock", lastBlock.getStringId());
    response.addProperty("lastBlockTimestamp", lastBlock.getTimestamp());
    response.addProperty(CUMULATIVE_DIFFICULTY_RESPONSE, lastBlock.getCumulativeDifficulty().toString());
    response.addProperty(AVERAGE_COMMITMENT_NQT_RESPONSE, lastBlock.getAverageCommitment());
    response.addProperty("numberOfBlocks", lastBlock.getHeight() + 1);
    Peer lastBlockchainFeeder = blockchainProcessor.getLastBlockchainFeeder();
    response.addProperty("lastBlockchainFeeder", lastBlockchainFeeder == null ? null : lastBlockchainFeeder.getAnnouncedAddress());
    response.addProperty("lastBlockchainFeederHeight", blockchainProcessor.getLastBlockchainFeederHeight());
    response.addProperty("isScanning", blockchainProcessor.isScanning());
    return response;
  }

}

File: src/main/kotlin/brs/http/GetBlockchainStatus.kt
package brs.http

import brs.Blockchain
import brs.BlockchainProcessor
import brs.Burst
import brs.http.common.ResultFields.TIME_RESPONSE
import brs.services.TimeService
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import javax.servlet.http.HttpServletRequest

internal class GetBlockchainStatus(private val blockchainProcessor: BlockchainProcessor, private val blockchain: Blockchain, private val timeService: TimeService) : APIServlet.JsonRequestHandler(arrayOf(APITag.BLOCKS, APITag.INFO)) {
    override fun processRequest(request: HttpServletRequest): JsonElement {
        val response = JsonObject()
        response.addProperty("application", Burst.APPLICATION)
        response.addProperty("version", Burst.VERSION.toString())
        response.addProperty(TIME_RESPONSE, timeService.epochTime)
        val lastBlock = blockchain.lastBlock
        response.addProperty("lastBlock", lastBlock.stringId)
        response.addProperty("cumulativeDifficulty", lastBlock.cumulativeDifficulty.toString())
        response.addProperty("numberOfBlocks", lastBlock.height + 1)
        val lastBlockchainFeeder = blockchainProcessor.lastBlockchainFeeder
        response.addProperty("lastBlockchainFeeder", lastBlockchainFeeder?.announcedAddress)
        response.addProperty("lastBlockchainFeederHeight", blockchainProcessor.lastBlockchainFeederHeight)
        return response
    }
}
